keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py

```diff
@@ -0,0 +1,199 @@
+      max_points_per_bounding_boxes: A int scalar sets the max number of points in a bounding box. All the object point clouds will be padded or trimmed to the same shape, where the number of points dimension is max_points_per_bounding_boxes.
+    """
+
+    def __init__(
+        self,
+        label_index=None,
+        min_points_per_bounding_boxes=0,
+        max_points_per_bounding_boxes=2000,
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+
+        if label_index and label_index < 0:
+            raise ValueError("label_index must be >=0 or None.")
+        if min_points_per_bounding_boxes < 0:
+            raise ValueError("min_points_per_bounding_boxes must be >=0.")
+        if max_points_per_bounding_boxes < 0:
+            raise ValueError("max_points_per_bounding_boxes must be >=0.")
+        if min_points_per_bounding_boxes > max_points_per_bounding_boxes:
+            raise ValueError(
+                "max_paste_bounding_boxes must be >= min_points_per_bounding_boxes."
+            )
+
+        self._label_index = label_index
+        self._min_points_per_bounding_boxes = min_points_per_bounding_boxes
+        self._max_points_per_bounding_boxes = max_points_per_bounding_boxes
+        self._iou_3d = iou_3d.IoU3D
+
+    def augment_point_clouds_bounding_boxes(
+        self, point_clouds, bounding_boxes, **kwargs
+    ):
+        if self._label_index:
+            bounding_boxes_mask = tf.math.equal(
+                bounding_boxes[0, :, 7], self._label_index
+            )
+            object_bounding_boxes = tf.boolean_mask(
+                bounding_boxes, bounding_boxes_mask, axis=1
+            )
+        else:
+            bounding_boxes_mask = ~tf.math.equal(bounding_boxes[0, :, 7], 0.0)
+            object_bounding_boxes = tf.boolean_mask(
+                bounding_boxes, bounding_boxes_mask, axis=1
+            )
+
+        points_in_bounding_boxes = is_within_box3d(
+            point_clouds, object_bounding_boxes[:, :, :7]
+        )
+        # Filter bounding boxes using the current frame.
+        min_points_filter = (
+            tf.reduce_sum(tf.cast(points_in_bounding_boxes[0], dtype=tf.int32), axis=0)
+            >= self._min_points_per_bounding_boxes
+        )
+
+        object_bounding_boxes = tf.boolean_mask(
+            object_bounding_boxes, min_points_filter, axis=1
+        )
+
+        points_in_bounding_boxes = tf.boolean_mask(
+            points_in_bounding_boxes, min_points_filter, axis=2
+        )
+        # [num of frames, num of boxes, num of points].
+        points_in_bounding_boxes = tf.transpose(points_in_bounding_boxes, [0, 2, 1])
+        points_in_bounding_boxes = tf.cast(points_in_bounding_boxes, tf.int32)
+        sort_valid_index = tf.argsort(
+            points_in_bounding_boxes, axis=-1, direction="DESCENDING"
+        )
+        sort_valid_mask = tf.gather(
```

Comment:
for padding -- shouldn't it pad with all zeros, instead of padding with points that are not within the box?

Reply:
Yes, padded object_point_clouds is replaced with zero before return, i.e.,

```
        object_point_clouds = tf.where(
            sort_valid_mask[:, :, :, tf.newaxis] > 0, object_point_clouds, 0.0
        )
```